### PR TITLE
Update docs for tooltip PR

### LINKED ISSUE #12168: Update docs for new Tooltip functionalities
Based on #12154, we need to update the docs. From a first glance, this should encompass at least some updates to the User Guide (including examples) and probably a stand-alone example as well.

With Tooltips now being a first class model, we could even think about creating a dedicated tooltips sub-chapter under 'Making interactions' (or somewhere else)!

### PR DESCRIPTION
**Pull request by tcmetzger**
Following https://github.com/bokeh/bokeh/pull/12154, we need to update the docs. I started this WIP PR to track the necessary docs updates. fixes #12168.

### CODE REVIEW

**Comment by tcmetzger** [comment]
@mattpap I'm working on the user guide entry for `BuiltinIcon` and am wondering where to point users to find a full list of Bokeh's built-in icons. The definition of `icon_name` in `icons.py` seems to indicate that this list might be the `ToolIcon` enumeration in `bokeh.core.enums`:

  ```
  icon_name = Required(Either(Enum(ToolIcon), String), help="""
    The name of a built-in icon to use.
    """)
```

But If I use `BuiltinIcon("save", size="1.2em", color="white")`, for example, I just get a box instead of the icon. What would be a good place to point users to to get a list of icon names that can get used with `BuiltinIcon`?

**Comment by mattpap** [comment]
> (...) am wondering where to point users to find a full list of Bokeh's built-in icons.

This is frankly not finished yet. Tool icons are still a good reference point, but only for historical reasons, as tools were there only place where we used icons. For now I would just point to tool icons and wait for the complete implementation.

**Comment by tcmetzger** [comment]
@mattpap Do I see correctly that we currently support the following built-in icons:
    * ``"help"``
    * ``"question-mark"``
    * ``"settings"``
    * ``"x"``
Do you plan on adding more built-in icons (or adding the ToolIcon icons) before the 3.0 release?
Either way, I'd say we include a list of the supported icon names in the helptext in icons.py, so it can be updated whenever the list of supported icons changes!

File: bokeh/models/ui/__init__.py [hunk 1]
```diff
@@ -26,12 +25,14 @@
     inspector,
     menus,
     tooltips,
+    ui_element,
```
Comment (tcmetzger) [inline bokeh/models/ui/__init__.py:28]: I needed to add this import to have the `UIElement` class included in the ref guide as well

File: bokeh/models/ui/ui_element.py [hunk 2]
```diff
@@ -43,7 +43,8 @@
 
 @abstract
 class UIElement(Model):
-    """ Base class for user interface elements. """
+    """ Base class for user interface elements.
+    """
```
Comment (tcmetzger) [inline bokeh/models/ui/ui_element.py:47]: I don't quite know why, but without this line break in the docstring, Sphinx throws an error: `docstring of bokeh.models.ui.ui_element.UIElement:3:Unexpected indentation.`

**Comment by tcmetzger** [comment]
@mattpap A question about the `InputWidget` object's `description` property: 

In `inputs.py`:90, this property is defined as follows:

`description = Nullable(Either(String, Instance(Tooltip)), default=None, help="""`

This seems to imply that I should be able to pass both a `Tooltip` object and a simple string. Based on this, I would expect the following to work:

```
from bokeh.io import show
from bokeh.models import ColorPicker
color_picker = ColorPicker(color="red", title="Choose color:", description="plain text string")
show(color_picker)
```

However, when running this, the question mark symbol is displayed but I don't get a tooltip popup!

Everything works as expected when I use a Tooltip object instead of the plain text string.

Should we change the definition of `description` to something like the following?

`description = Nullable(Instance(Tooltip)), default=None, help="""`

**Comment by tcmetzger** [comment]
@mattpap I'm looking at the `Tooltip` object and was wondering why the default value for the tooltip's `position` property is `None`. This essentially seems to mean that when using the default value for Tooltip, the tooltip does not get displayed.

A simple example like this doesn't display a tooltip:

```
from bokeh.models import Tooltip, NumericInput
from bokeh.io import show

numeric_input = NumericInput(
    title="Choose value:",
    value=1, low=0, high=100,
    description=Tooltip(content="Enter a number between 0 and 100")
    )

show(numeric_input)
```

It seems like I have to use `description=Tooltip(content="Enter a number between 0 and 100", position="right")` (with a value passed tp `position`) to get a tooltip.

Wouldn't it be more user-friendly to allow `description=Tooltip(content="Enter a number between 0 and 100")` to work with a default value for `position`, instead of being required to basically always have a value for the `position` property?

**Comment by mattpap** [comment]
I suppose `Tooltip.position` (and the general API) requires a bit more refinement.

**Comment by tcmetzger** [comment]
@mattpap another small observation: I found that all objects in "inputs" (https://docs.bokeh.org/en/latest/docs/reference/models/widgets/inputs.html) now support the `description` parameter - except for the `Switch` object (which is based on `Widget` instead of `InputWidget`, unlike all other objects in this list).

Would it be possible to add support for `description` to the `Switch` widget as well (e.g. make it a descendant of `InputWidget`, like all other input widgets)? This way we wouldn't have to say something like "all input widgets except for `Switch` support tooltips with the `description` parameter").

**Comment by tcmetzger** [comment]
@bryevdv I'm now working on an explanation of how to use the Tooltip's ``target`` property to attach a tooltip to an arbitrary UI element. I have created an example that showcases both selecting by Bokeh model (for ``text_input``) and a
`ByClass` object (for ``div``):

```
from bokeh.document import Document
from bokeh.embed import file_html
from bokeh.models import ByClass, Div, TextInput, Tooltip
from bokeh.util.browser import view

# define text input widget and corresponding tooltip
text_input = TextInput(value="Text input")
tooltip_1 = Tooltip(
    content="Enter text here",
    position="right",
    attachment="right",
    target=text_input,
    closable=True,
    visible=True,
)

# define Div widget and corresponding tooltip
div = Div(
    text="A Div using a specific CSS class (<code>my-css-class</code>)",
    css_classes=["my-css-class"],
)
tooltip_2 = Tooltip(
    content="More about this Div",
    position="right",
    attachment="right",
    target=ByClass("my-css-class"),  # Selecting the Div by its CSS class
    closable=True,
    visible=True,
)

# create new document and add objects
doc = Document()
doc.add_root(text_input)
doc.add_root(tooltip_1)
doc.add_root(div)
doc.add_root(tooltip_2)

if __name__ == "__main__":
    doc.validate()
    filename = "tooltips_testing.html"
    with open(filename, "w") as f:
        f.write(file_html(doc, "Tooltips"))
    print("Wrote %s" % filename)
    view(filename)
```

When I put this code in a file and run this file as a standalone script, things work as expected. But when I try to use this code in an RST file with a `bokeh-plot` directive, I get an error because I'm adding too many things to the document root (`bokeh-plot:: directive expects a single Document root, got 4`).

What would be a good way to modify this example to work with the `bokeh-plot` directive?

**Comment by bryevdv** [comment]
@tcmetzger Is the intent to show how to add Bokeh tooltips to non-Bokeh DOM elements, except the example uses Bokeh DOM elements because that is simpler than embedding in a Jinja template? Otherwise I would say I don't understand why this would be preferable to `description=Tooltip(...)`. 

In case this is an advanced example, and I don't think there is a simple way to make it work in the docs. I would say add an example in the repo, but don't worry about the docs or now.

**Comment by tcmetzger** [comment]
@bryevdv Yes, that was exactly the idea. You are right, this is probably too complex for the user guide, a standalone example does make more sense!

File: sphinx/source/docs/user_guide/interaction/js_callbacks.rst [hunk 3]
```diff
@@ -72,33 +121,35 @@ any Bokeh model, using the ``js_on_change`` method of Bokeh models:
...
-patched or streamed to.
+The first parameter to ``js_on_change`` is actually the name of a BokehJS event.
+The full format for a property change event is, for example, ``"change:start"``.
+However, Bokeh will automatically convert any property name into one of these
+BokehJS change events for you.
+
+Some Bokeh models have additional, specialized events. For example, the
+:class:`~bokeh.models.ColumnDataSource` model also supports ``"patch"`` and
+``"stream"`` events. You can use these events to trigger ``CustomJS`` callbacks
+whenever the data source is patched or streamed to.
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/interaction/js_callbacks.rst:132]: I wonder if this is too much information, we don't demonstrate the `"change:foo"` usage anywhere really, and I am not really sure we should advertise it.

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
Looks great @tcmetzger just one comment

**Pull request merged**